### ISSUE TITLE
[IMPROVEMENT] Thoughts on DDI Data Loader

### ISSUE THREAD

**Issue by TheCedarPrince** [JuliaHealth/IPUMS.jl#22]
Hey @00krishna,

These are my thoughts on your recent implementation for #21! As I mentioned over Slack, the functionality is sufficient enough that I went ahead and merged the PR, but as we discussed, there does seem to be some places for improvements. I am going to copy some of our discussion to here first so as to not lose thoughts as well as add in my own thoughts and comments along the way.

### Basic Benchmarking

For a dataset with about 70 million elements, I see this time and allocation count:

```julia-repl
julia> @btime load_ipums_extract(ddi, "usa_00001.dat");
  130.577 s (475777044 allocations: 15.20 GiB) 
```

For a dataset with only 60,000 elements, I see this time and allocation count:

```julia-repl
julia> @btime load_ipums_extract(ddi, "cps_00157.dat");
  118.328 ms (518521 allocations: 17.71 MiB)
```

Interestingly, when I tried using `ipumsr` to do this, the loading for both files was nearly instantaneous. I did some digging and found that [hipread](https://github.com/cran/hipread) seems to be doing a lot of the heavy-lifting within `ipumsr` to do the parsing of this file.

### Thoughts on Slowdowns

Per Krishna:

> Most of the allocations must be happening in this part of the code, since this is where the data is getting loaded.
>```julia
>   for line in eachline(extract_filepath)
>        data = map((x,p,d) -> _parse_data(x, p, d), 
>                        [strip(line[r]) for r in range_vec], 
>                        [eltype(a) for a in dtype_vec],
>                        [d for d in dcml_vec])
>        push!(df, data)
>    end
> ```

After doing some further analysis using Profile.jl and PProf.jl, I was able to find the following flamegraph profiling on allocations in the code:

![image](https://github.com/JuliaHealth/IPUMS.jl/assets/29561456/4ebb455c-4a4c-4c68-a60e-c737330137d6)

Although the image is not so great, one can see that a majority of the allocations occurs in the map function call as you were imagining, Krishna. Here's the code I ran to generate this flamegraph:

```julia
using IPUMS
using Profile
using PProf

ddi = parse_ddi("cps_00157.xml");

Profile.Allocs.clear()
Profile.Allocs.@profile load_ipums_extract(ddi, "cps_00157.dat");
PProf.Allocs.pprof()
```

I attached the allocation profile file here as well.

[alloc-profile.pb.gz](https://github.com/JuliaHealth/IPUMS.jl/files/15215650/alloc-profile.pb.gz)

I also did a quick analysis of the computation. It would appear that much of the time the function is running, it is trying its best at type inference. This flame graph is a bit inscrutable but I figured I would include it here as well:

![image](https://github.com/JuliaHealth/IPUMS.jl/assets/29561456/8bc090d2-6bc6-442c-a7d2-ff876228243e)

### Concluding Thoughts

In short, it seems like there is a huge opportunity to optimize the loading functionality to make it more straightforwardly take advantage of the fixed width files. Although, I am not entirely sure how to do that yet. I'd be curious your thoughts @00krishna.

Otherwise, great stuff!

**Comment by 00krishna** [comment]
So I have created PR https://github.com/JuliaHealth/IPUMS.jl/pull/30 to help improve on the performance. @TheCedarPrince is reviewing this. We will work on further improvements as well.

**Comment by TheCedarPrince** [comment]
That was merged @00krishna -- thanks for adding this in. Are you still able to post that minimal working example to Discourse? I'd be curious to see how we could keep hacking at it someday to make it as fast as tidyr's reader.

**Comment by TheCedarPrince** [comment]
Hey @riteshbhirud ! Just some additional context on some of the initial implementations: [Regex assistance converting from R to Julia](https://discourse.julialang.org/t/regex-assistance-converting-from-r-to-julia/112852)

**Comment by TheCedarPrince** [comment]
OK! After strong efforts from @riteshbhirud , the parser is now tremendously faster (via #40)! A few interesting observations however:

For small files, we are currently beating `ipumsr`:

```R
> ddi <- read_ipums_ddi("test/testdata/cps_00157.xml")
> data_file <- "test/testdata/cps_00157.dat"
> bench::mark(                                        
  read_ipums_micro(ddi, data_file = data_file)
)
Use of data from IPUMS CPS is subject to conditions including that users should cite the data appropriately. Use command `ipums_condition
s()` for more details.
# A tibble: 1 × 13
  expression      min median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc total_time
  <bch:expr>   <bch:> <bch:>     <dbl> <bch:byt>    <dbl> <int> <dbl>   <bch:tm>
1 read_ipums_… 7.49ms 7.93ms      122.     925KB     27.9    48    11      394ms
# ℹ 4 more variables: result <list>, memory <list>, time <list>, gc <list>
```

Versus our implementation: 

```julia
julia> @benchmark load_ipums_extract($ddi, $("test/testdata/cps_00157.dat"))
BenchmarkTools.Trial: 890 samples with 1 evaluation per sample.
 Range (min … max):  5.014 ms … 34.428 ms  ┊ GC (min … max): 0.00% … 83.09%
 Time  (median):     5.490 ms              ┊ GC (median):    0.00%
 Time  (mean ± σ):   5.617 ms ±  1.264 ms  ┊ GC (mean ± σ):  1.22% ±  4.64%

     ▅▄▅▅▂█▃                                                  
  ▂▃▄███████▇▆▅▄▄▃▂▂▂▃▂▃▂▂▁▂▂▁▂▂▁▁▁▂▁▁▁▂▁▁▁▁▁▂▁▁▁▁▂▂▁▁▁▂▂▁▁▂ ▃
  5.01 ms        Histogram: frequency by time        8.63 ms <

 Memory estimate: 1.24 MiB, allocs estimate: 15544.
```

**About a 150% improvement! That is pretty incredible! 🥳** 

However, for large files, we seem to have a kind of performance hit where `ipumsr` completely smokes this new parser; here is the benchmark on a ~1GB file:

```R
> cps_ddi_file <- read_ipums_ddi("file_downloads/cps_00001.xml")
> data_file <- "file_downloads/cps_00001.dat"
> bench::mark(                                                  
  read_ipums_micro(ddi, data_file = data_file)
)
Use of data from IPUMS CPS is subject to conditions including that users should cite the data appropriately. Use command `ipums_condition
s()` for more details.
|=================================================================| 100% 1182 MB
# A tibble: 1 × 13
  expression      min median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc total_time
  <bch:expr>    <bch> <bch:>     <dbl> <bch:byt>    <dbl> <int> <dbl>   <bch:tm>
1 read_ipums_m…  1.8s   1.8s     0.555     897MB     1.66     1     3       1.8s
# ℹ 4 more variables: result <list>, memory <list>, time <list>, gc <list>
Warning message:
Some expressions had a GC in every iteration; so filtering is disabled. 
```

versus the new Julia parser

```julia-repl
julia> @benchmark load_ipums_extract($ddi, $("file_downloads/cps_00001.dat"))
BenchmarkTools.Trial: 1 sample with 1 evaluation per sample.
 Single result which took 21.614 s (4.28% GC) to evaluate,
 with a memory estimate of 3.53 GiB, over 16153927 allocations.
```

Where now, the Julia parser is getting beaten by ~1000%. I am not sure what is not scaling here or if there is some sort of bifurcation drop-off where we should switch to some other routine, but I wanted to list this as a general note while we keep working on things across IPUMS.jl!
